File: word-break-exhaustive.cpp
// Time Complexity : O(n^n)
// Space Complexity : O(nL)
// Did this code successfully run on Leetcode : NO (Time Limit Exceeded)

#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

//exhaustive recursive
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> set;
        for(auto it: wordDict){
            set.insert(it);
        }
        return helper(s, set);
    }
private:
    bool helper(string curr, unordered_set<string> set){
        //base
        if(curr.size() == 0) return true;
        //logic
        for(int i = 0; i < curr.size(); i++){
            string sub = curr.substr(0, i + 1);
            if(set.find(sub) != set.end() && helper(curr.substr(i + 1), set)){
                return true;
            }
        }
        return false;
    }
};

File: word_break_exhaustive.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        if len(s) == 0 or len(wordDict) == 0:
            return true
        hashset = set()
        for word in wordDict:
            hashset.add(word)

        return self.helper(s, hashset)

    def helper(self, s, hashset):
        # base
        if len(s) == 0:
            return True

        # logic
        for i in range(0, len(s)):
            substring = s[:i+1]
            if substring in hashset and self.helper(s[i+1:], hashset):
                return True

        return False
